graph/preprocessor.go

```diff
@@ -0,0 +1,291 @@
+		}
+	}
+	return nil
+}
+
+// preprocessString handles the preprocessing (string replacement and resolution)
+// of all aliases in an input yaml (passed in as a string). The resolved aliases are
+// defined in the input alias file.
+func preprocessString(alias *Alias, str string) (string, bool, error) {
+	// alias.loadGlobalDefinitions TODO?
+
+	// Load Remote/Local alias definitions
+	if externalDefinitionErr := alias.loadExternalAlias(); externalDefinitionErr != nil {
+		return "", false, externalDefinitionErr
+	}
+	// Validate alias definitions
+	if improperFormatErr := alias.resolveMapAndValidate(); improperFormatErr != nil {
+		return "", false, improperFormatErr
+	}
+	var out strings.Builder
+	var command strings.Builder
+	ongoingCmd := false
+	changed := false
+
+	// Search and replace all strings with the directive
+	for _, char := range str {
+		if ongoingCmd {
+			if char == alias.directive && command.Len() == 0 { // Escape Character Triggered
+				out.WriteRune(alias.directive)
+				ongoingCmd = false
+
+			} else if !isAlphanumeric(char) { // Delineates the end of an alias
+				resolvedCommand, commandPresent := alias.AliasMap[command.String()]
+				// If command is not found we assume this to be the expect item itself.
+				if !commandPresent {
+					out.WriteString(string(alias.directive) + command.String())
+					command.Reset()
+				} else {
+					out.WriteString(resolvedCommand)
+					changed = true
+					if char != alias.directive {
+						ongoingCmd = false
+						out.WriteRune(char)
+					}
+					command.Reset()
+				}
+			} else {
+				command.WriteRune(char)
+			}
+		} else if char == alias.directive {
+			ongoingCmd = true
+		} else {
+			out.WriteRune(char)
+		}
+	}
+
+	return out.String(), changed, nil
+}
+
+// preprocessBytes handles byte encoded data that can be parsed through pre processing
+func preprocessBytes(data []byte) ([]byte, Alias, bool, error) {
+	type Wrapper struct {
+		Alias Alias `yaml:"alias,omitempty"`
+	}
+	wrap := &Wrapper{}
+	aliasData, remainingData, err := basicAliasSeparation(data)
+	if errUnMarshal := yaml.Unmarshal(aliasData, wrap); errUnMarshal != nil {
+		return data, Alias{}, false, errUnMarshal
+	}
+
+	alias := &wrap.Alias
+	if alias.AliasMap == nil {
+		// Nothing to change
+		if alias.AliasSrc == nil {
+			return data, *alias, false, nil
+		}
+		// Alias Src defined. Guarantees alias map can be populated
+		alias.AliasMap = make(map[string]string)
+	}
+
+	// Search and Replace
+	str := string(remainingData)
+	parsedStr, changed, err := preprocessString(alias, str)
+
+	return []byte(parsedStr), *alias, changed, err
+}
+
+// processSteps Will resolve image names in steps that are aliased without using $.
+// Invoked after resolving $
+func processSteps(alias *Alias, task *Task) {
+	for i, step := range task.Steps {
+		parts := strings.Split(step.Cmd, " ")
+		if _, ok := alias.AliasMap[parts[0]]; ok {
+			// Image name should always go first
+			parts[0] = alias.AliasMap[parts[0]]
```

Comment:
Cannot assume every command will have an alias as the first argument. For the following command:

```
- cmd: mcr.microsoft.com/acb exec ...
```

`alias.AliasMap["mcr.microsoft.com/acb"] does not exist.

```
if val, ok := alias.AliasMap[parts[0]]; ok {
    parts[0] = alias.AliasMap[parts[0]]
    task.Steps[i].Cmd = strings.Join(parts, " ")
}
```

Reply:
I am a little confused by the first assumption. From my understanding, I am only replacing the value if the first argument can be detected to be an alias through its presence in the map. Are you speaking about the assumption that it will be the first item? Eric told me that was a fair assumption as while this may not always be true, its rare for people to do something different and we wish to encourage the behavior where the image name comes first (or an alias in this scenario). If the value has no alias , in this case, its just kept as it was before.